# Welcome to your first assignment!

Instructions are below. Please give this a try, and look in the solutions folder if you get stuck (or feel free to ask me!)

<table style="margin: 0; text-align: left;">
    <tr>
        <td style="width: 150px; height: 150px; vertical-align: middle;">
            <img src="../resources.jpg" width="150" height="150" style="display: block;" />
        </td>
        <td>
            <h2 style="color:#f71;">Just before we get to the assignment --</h2>
            <span style="color:#f71;">I thought I'd take a second to point you at this page of useful resources for the course. This includes links to all the slides.<br/>
            <a href="https://edwarddonner.com/2024/11/13/llm-engineering-resources/">https://edwarddonner.com/2024/11/13/llm-engineering-resources/</a><br/>
            Please keep this bookmarked, and I'll continue to add more useful links there over time.
            </span>
        </td>
    </tr>
</table>

# HOMEWORK EXERCISE ASSIGNMENT

Upgrade the day 1 project to summarize a webpage to use an Open Source model running locally via Ollama rather than OpenAI

You'll be able to use this technique for all subsequent projects if you'd prefer not to use paid APIs.

**Benefits:**
1. No API charges - open-source
2. Data doesn't leave your box

**Disadvantages:**
1. Significantly less power than Frontier Model

## Recap on installation of Ollama

Simply visit [ollama.com](https://ollama.com) and install!

Once complete, the ollama server should already be running locally.  
If you visit:  
[http://localhost:11434/](http://localhost:11434/)

You should see the message `Ollama is running`.  

If not, bring up a new Terminal (Mac) or Powershell (Windows) and enter `ollama serve`  
And in another Terminal (Mac) or Powershell (Windows), enter `ollama pull llama3.2`  
Then try [http://localhost:11434/](http://localhost:11434/) again.

If Ollama is slow on your machine, try using `llama3.2:1b` as an alternative. Run `ollama pull llama3.2:1b` from a Terminal or Powershell, and change the code below from `MODEL = "llama3.2"` to `MODEL = "llama3.2:1b"`

In [1]:
# imports

import requests
from bs4 import BeautifulSoup
from IPython.display import Markdown, display

In [2]:
# Constants

OLLAMA_API = "http://localhost:11434/api/chat"
HEADERS = {"Content-Type": "application/json"}
MODEL = "llama3.2"

In [3]:
# Create a messages list using the same format that we used for OpenAI

messages = [
    {"role": "user", "content": "Describe some of the business applications of Generative AI"}
]

In [4]:
payload = {
        "model": MODEL,
        "messages": messages,
        "stream": False
    }

In [5]:
# Let's just make sure the model is loaded

!ollama pull llama3.2

pulling manifest ⠋ pulling manifest ⠙ pulling manifest 
pulling dde5aa3fc5ff... 100% ▕████████████████▏ 2.0 GB                         
pulling 966de95ca8a6... 100% ▕████████████████▏ 1.4 KB                         
pulling fcc5a6bec9da... 100% ▕████████████████▏ 7.7 KB                         
pulling a70ff7e570d9... 100% ▕████████████████▏ 6.0 KB                         
pulling 56bb8bd477a5... 100% ▕████████████████▏   96 B                         
pulling 34bb5ab01051... 100% ▕████████████████▏  561 B                         pulling manifest 
pulling dde5aa3fc5ff... 100% ▕████████████████▏ 2.0 GB                         
pulling 966de95ca8a6... 100% ▕████████████████▏ 1.4 KB                         
pulling fcc5a6bec9da... 100% ▕████████████████▏ 7.7 KB                         
pulling a70ff7e570d9... 100% ▕████████████████▏ 6.0 KB                         
pulling 56bb8bd477a5... 100% ▕████████████████▏   96 B                         
pulling 34bb5ab01051... 100% ▕████████████████▏

In [6]:
# If this doesn't work for any reason, try the 2 versions in the following cells
# And double check the instructions in the 'Recap on installation of Ollama' at the top of this lab
# And if none of that works - contact me!

response = requests.post(OLLAMA_API, json=payload, headers=HEADERS)
print(response.json()['message']['content'])

Generative AI has numerous business applications across various industries, including:

1. **Content Generation**: AI-powered tools can generate high-quality content such as articles, social media posts, product descriptions, and more. This can help businesses save time and resources while maintaining consistency in their content.
2. **Marketing and Advertising**: Generative AI can be used to create personalized advertisements, product recommendations, and even entire marketing campaigns. It can also help analyze customer data and behavior to inform targeted advertising strategies.
3. **Product Design and Development**: AI-powered generative design tools can quickly create prototypes, 3D models, and even entire product lines. This can accelerate the product development process and reduce costs associated with traditional design methods.
4. **Customer Service and Support**: Generative AI-powered chatbots and virtual assistants can provide 24/7 customer support, helping businesses respon

# Introducing the ollama package

And now we'll do the same thing, but using the elegant ollama python package instead of a direct HTTP call.

Under the hood, it's making the same call as above to the ollama server running at localhost:11434

In [7]:
import ollama

response = ollama.chat(model=MODEL, messages=messages)
print(response['message']['content'])

Generative AI has numerous business applications across various industries. Here are some examples:

1. **Content Creation**: Generative AI can be used to generate high-quality content such as articles, social media posts, product descriptions, and more. This can help businesses save time and resources on content creation.
2. **Product Design**: Generative AI can be used to design new products, such as furniture, clothing, or electronics. It can also be used to generate 3D models and prototypes.
3. **Image and Video Generation**: Generative AI can be used to generate images and videos that mimic real-world scenarios, such as product demonstrations, customer testimonials, or marketing campaigns.
4. **Chatbots and Virtual Assistants**: Generative AI can be used to create chatbots and virtual assistants that can understand and respond to user queries in a more natural and human-like way.
5. **Predictive Analytics**: Generative AI can be used to analyze large datasets and predict future tr

## Alternative approach - using OpenAI python library to connect to Ollama

In [8]:
# There's actually an alternative approach that some people might prefer
# You can use the OpenAI client python library to call Ollama:

from openai import OpenAI
ollama_via_openai = OpenAI(base_url='http://localhost:11434/v1', api_key='ollama')

response = ollama_via_openai.chat.completions.create(
    model=MODEL,
    messages=messages
)

print(response.choices[0].message.content)

Generative AI (Artificial Intelligence) is transforming various industries and has numerous business applications. Here are some examples:

1. **Content Creation**: Generative AI can create high-quality, unique content such as:
	* Text: articles, blog posts, social media posts, and product descriptions.
	* Images: photos, videos, and even personalized product designs.
	* Music: original compositions and beats.
2. **Marketing Automation**: Generative AI can help automate marketing processes by creating:
	* Personalized advertising content
	* Email campaigns
	* Social media posts
	* Sales scripts
3. **Customer Service**: Generative AI-powered chatbots can provide:
	* 24/7 customer support
	* Customer inquiries and feedback analysis
	* Basic troubleshooting and issue resolution
4. **Product Design and Prototyping**: Generative AI can help designers create:
	* New product designs
	* Prototypes for testing and iteration
	* Optimized manufacturing processes
5. **Data Analysis and Visualizati

# NOW the exercise for you

Take the code from day1 and incorporate it here, to build a website summarizer that uses Llama 3.2 running locally instead of OpenAI; use either of the above approaches.

In [9]:
# updating Website class to handle dynamic websites
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import time

class Website:
    def __init__(self, url):
        """
        Create this Website object from the given url using Selenium and BeautifulSoup
        """
        self.url = url

        # Set up Selenium with Chrome using WebDriver Manager
        chrome_options = Options()
        chrome_options.add_argument("--headless")  # Run in headless mode
        service = Service(ChromeDriverManager().install())
        driver = webdriver.Chrome(service=service, options=chrome_options)

        try:
            # Open the URL
            driver.get(url)

            # Wait for the dynamic content to load
            WebDriverWait(driver, 10).until(
                EC.presence_of_element_located((By.TAG_NAME, "body"))
            )

            # Get page source and parse with BeautifulSoup
            soup = BeautifulSoup(driver.page_source, 'html.parser')
            self.title = soup.title.string if soup.title else "No title found"
            for irrelevant in soup.body(["script", "style", "img", "input"]):
                irrelevant.decompose()
            self.text = soup.body.get_text(separator="\n", strip=True)
        
        finally:
            # Close the browser
            driver.quit()

In [12]:
# Step 0: Scrape website
q1 = Website("https://www.nba.com/game/nyk-vs-was-0022400429")

# Step 1: Create your prompts

system_prompt = """
    You are a meticulous and detailed assistant that summarizes that provides detailed nba game summaries in a consumable format.
    You receive the content of the website and respond in organized markdown.
    Ignore text that doesn't belong to the website, like navigation related text.
"""
user_prompt = f"""
    Can you summarize the game? pleae provide the key points?
    Game Information: {q1.title}
    Game Summary: {q1.text}
"""

# Step 2: Make the messages list
messages = [
    {"role": "system", "content": system_prompt},
    {"role": "user", "content": user_prompt}
] # fill this in

# Step 3: Call Ollama
response = ollama.chat(model=MODEL, messages=messages)

# Step 4: print the result
print(response['message']['content'])

**Game Summary: New York Knicks vs Washington Wizards**

### Game Information

* Date: December 28, 2024
* Location: Capital One Arena, Washington D.C.
* Attendance: 20,385
* Officials: Josh Tiven, Karl Lane, Tyler Ricks

### Key Points

#### Game Details

* Final Score: New York Knicks 136, Washington Wizards 132 (OT)
* Game Type: Overtime
* Location: Capital One Arena, Washington D.C.

#### Player Performances

* **Jalen Brunson** (NYK): 55 points, 9 assists, 2 jump shots and 5 free throws in OT
* **Karl-Anthony Towns** (NYK): 30 points, 14 rebounds
* **Justin Champagnie** (NYK): 31 points, 10 rebounds
* **Malcolm Brogdon** (WAS): 22 points

#### Game Recap

* The Knicks extended their season-long winning streak to seven games.
* The Wizards led for most of the game but committed 17 turnovers, which resulted in 26 New York points.
* Brunson scored nine of his season-high 55 points in overtime and hit five free throws in the extra period.

### Box Score

| Team | Q1 | Q2 | Q3 | Q4 | O

### Game Information

* Date: December 28, 2024
* Location: Capital One Arena, Washington D.C.
* Attendance: 20,385
* Officials: Josh Tiven, Karl Lane, Tyler Ricks

### Key Points

#### Game Details

* Final Score: New York Knicks 136, Washington Wizards 132 (OT)
* Game Type: Overtime
* Location: Capital One Arena, Washington D.C.

#### Player Performances

* **Jalen Brunson** (NYK): 55 points, 9 assists, 2 jump shots and 5 free throws in OT
* **Karl-Anthony Towns** (NYK): 30 points, 14 rebounds
* **Justin Champagnie** (NYK): 31 points, 10 rebounds
* **Malcolm Brogdon** (WAS): 22 points

#### Game Recap

* The Knicks extended their season-long winning streak to seven games.
* The Wizards led for most of the game but committed 17 turnovers, which resulted in 26 New York points.
* Brunson scored nine of his season-high 55 points in overtime and hit five free throws in the extra period.

### Box Score

| Team | Q1 | Q2 | Q3 | Q4 | OT | Total |
| --- | --- | --- | --- | --- | --- | --- |
| NYK | 32 | 28 | 30 | 29 | 17 | 136 |
| WAS | 33 | 27 | 38 | 21 | 13 | 132 |

### Lead Changes and Times Tied

* 25 lead changes
* 12 times tied

### Gamebook

Available upon request.

---

**Game Stats**

| Category | NYK | WAS |
| --- | --- | --- |
| FB PTS | 78 | 60 |
| BIG LD | 14 | 18 |
| BPTS | 6 | 11 |
| TREB | 8 | 32 |
| TOV | 6 | 32 |
| TTOV | 8 | 15 |
| POT | 26 | 5 |

---

**Referees**

* Josh Tiven
* Karl Lane
* Tyler Ricks